### PR TITLE
Leaflet vector layer

### PR DESCRIPTION
**Pull request by vlach1989**
- general VectorLayer component
- enable to render tile as canvas
- layer type vs. rendering technique

### CODE REVIEW

File: src/ReactLeafletMap/layers/CanvasVectorLayer/LeafletCanvasLayer.js [hunk 1]
```diff
@@ -45,7 +45,7 @@ const LeafletCanvasLayer = L.CanvasLayer.extend({
 			const self = this;
 
 			// TODO breakable loop?
-			this.features.forEach(feature => {
+			this.features?.forEach(feature => {
```
Comment (vdubr) [inline src/ReactLeafletMap/layers/CanvasVectorLayer/LeafletCanvasLayer.js:48]: Secure if features is null.

File: src/ReactLeafletMap/layers/IndexedVectorLayer/index.js [hunk 2]
```diff
@@ -46,6 +49,8 @@ class IndexedVectorLayer extends React.PureComponent {
     }
 
 	componentDidUpdate(prevProps, prevState, snapshot) {
+		// console.log("options", prevProps.options === this.props.options);
```
Comment (vdubr) [inline src/ReactLeafletMap/layers/IndexedVectorLayer/index.js:52]: remove console.logs ?

File: src/ReactLeafletMap/layers/TiledVectorLayer/Tile.js [hunk 3]
```diff
@@ -89,19 +91,31 @@ class Tile extends React.PureComponent {
 	}
 
 	render() {
-		const {tileKey, featureKeysGroupedByTileKey, ...props} = this.props;
+		const {tileKey, featureKeysGroupedByTileKey, component, ...props} = this.props;
```
Comment (vdubr) [inline src/ReactLeafletMap/layers/TiledVectorLayer/Tile.js:94]: Add component to propTypes as required?

File: src/ReactLeafletMap/layers/VectorLayer/index.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		onClick: PropTypes.func,
+		opacity: PropTypes.number,
+		options: PropTypes.object,
+		type: PropTypes.string,
+		view: PropTypes.object,
+		zoom: PropTypes.number,
+		zIndex: PropTypes.number
+	};
+
+	constructor(props) {
```
Comment (vdubr) [inline src/ReactLeafletMap/layers/VectorLayer/index.js:27]: it is not necessary, is not it?

File: src/ReactLeafletMap/layers/VectorLayer/index.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+		const renderAs = props.options?.renderAs;
+		let options = props.options;
+
+		if (renderAs) {
+			const boxRange  = props.view?.boxRange;
+			const renderAsData = _.find(renderAs, (renderAsItem) => {
+				const boxRangeRange = renderAsItem.boxRangeRange;
+
+				// Current boxRange is in defined range
+				return (boxRange > boxRangeRange[0] && boxRange <= boxRangeRange[1]) || (!boxRangeRange[0] && boxRange <= boxRangeRange[1]) || (boxRange > boxRangeRange[0] && !boxRangeRange[1]);
```
Comment (vdubr) [inline src/ReactLeafletMap/layers/VectorLayer/index.js:42]: Do we have some helper for it?

**Review by vdubr — APPROVED** [review]

**Review by tmdmc — APPROVED** [review]

File: src/ReactLeafletMap/layers/CanvasVectorLayer/LeafletCanvasLayer.js (hunk 1 above)
Comment (vlach1989) [reply src/ReactLeafletMap/layers/CanvasVectorLayer/LeafletCanvasLayer.js]: 4dd36b335958ec8c9d68eb519561ade5c80faa60

File: src/ReactLeafletMap/layers/IndexedVectorLayer/index.js (hunk 2 above)
Comment (vlach1989) [reply src/ReactLeafletMap/layers/IndexedVectorLayer/index.js]: 5c9fad407379c7e55257fd0481954cb4776bd499

File: src/ReactLeafletMap/layers/TiledVectorLayer/Tile.js (hunk 3 above)
Comment (vlach1989) [reply src/ReactLeafletMap/layers/TiledVectorLayer/Tile.js:98]: e0a7d3e2f03c00f2dfc62368f290ebebdc622270

File: src/ReactLeafletMap/layers/VectorLayer/index.js (hunk 4 above)
Comment (vlach1989) [reply src/ReactLeafletMap/layers/VectorLayer/index.js:27]: why?

File: src/ReactLeafletMap/layers/VectorLayer/index.js (hunk 5 above)
Comment (vlach1989) [reply src/ReactLeafletMap/layers/VectorLayer/index.js:42]: You mean for line 42?

File: src/ReactLeafletMap/layers/VectorLayer/index.js (hunk 4 above)
Comment (vdubr) [reply src/ReactLeafletMap/layers/VectorLayer/index.js:27]: https://reactjs.org/docs/react-component.html#constructor

File: src/ReactLeafletMap/layers/VectorLayer/index.js (hunk 5 above)
Comment (vdubr) [reply src/ReactLeafletMap/layers/VectorLayer/index.js:42]: yes

File: src/ReactLeafletMap/layers/VectorLayer/index.js (hunk 4 above)
Comment (vlach1989) [reply src/ReactLeafletMap/layers/VectorLayer/index.js]: 632951677d08608be18ef31770d28b59e3aa468d

File: src/ReactLeafletMap/layers/VectorLayer/index.js (hunk 5 above)
Comment (vlach1989) [reply src/ReactLeafletMap/layers/VectorLayer/index.js]: 55672d5f0a797875a6b75c07714f451d1ac8b6cb

**Review by vdubr — APPROVED** [review]

**Pull request merged**